akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala

```diff
@@ -793,28 +796,45 @@ final class Partition[T](val outputPorts: Int, val partitioner: T => Int, val ea
 
   override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
     new GraphStageLogic(shape) with InHandler {
+      lazy val decider = inheritedAttributes.mandatoryAttribute[SupervisionStrategy].decider
       private var outPendingElem: Any = null
       private var outPendingIdx: Int = _
       private var downstreamRunning = outputPorts
 
       def onPush() = {
         val elem = grab(in)
-        val idx = partitioner(elem)
-        if (idx < 0 || idx >= outputPorts) {
-          failStage(PartitionOutOfBoundsException(
-            s"partitioner must return an index in the range [0,${outputPorts - 1}]. returned: [$idx] for input [${elem.getClass.getName}]."))
-        } else if (!isClosed(out(idx))) {
-          if (isAvailable(out(idx))) {
-            push(out(idx), elem)
-            if (out.exists(isAvailable(_)))
-              pull(in)
-          } else {
-            outPendingElem = elem
-            outPendingIdx = idx
+        val idx =
+          try {
+            val i = partitioner(elem)
+            if (i < 0 || i >= outputPorts)
+              throw PartitionOutOfBoundsException(
+                s"partitioner must return an index in the range [0,${outputPorts - 1}]. returned: [$i] for input " +
+                s"[${elem.getClass.getName}].")
+            else i
+          } catch {
+            case NonFatal(ex) =>
+              decider(ex) match {
+                case Supervision.Stop ⇒ failStage(ex)
+                case Supervision.Restart ⇒ pull(in)
+                case Supervision.Resume ⇒ pull(in)
+              }
+              Int.MinValue
           }
 
-        } else if (out.exists(isAvailable(_)))
-          pull(in)
+        if (idx != Int.MinValue) {
+          if (!isClosed(out(idx))) {
+            if (isAvailable(out(idx))) {
+              push(out(idx), elem)
+              if (out.exists(isAvailable(_)))
+                pull(in)
+            } else {
+              outPendingElem = elem
+              outPendingIdx = idx
+            }
+
+          } else if (out.exists(isAvailable(_)))
+            pull(in)
+        }
```

Comment:
Can this be done without a 4-level if statement?

Reply:
I extracted the pull part, see d2bf559, but more than that I don't see how it can be simplified. The API style in graph stages is very (low level) imperative for performance reasons.